### ISSUE TITLE
`save_hyperparameters` does not save the selected` learning_rate` when `auto_lr_find` is used

### ISSUE THREAD

**Issue by iceboundflame** [Lightning-AI/pytorch-lightning#15928]
### Bug description

I am running my trainer with the auto_lr_find option, but the model's saved learning_rate reflects the original learning rate and not the final one selected by trainer.tune(). This is because save_hyperparameters() needs to be called at model construction, however it then cannot save the final selected learning rate. This feels like a shortcoming - is there a recommended workaround here? I didn't see one in the docs. Thanks!

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @awaelchli @borda @Blaizzy

**Comment by Palzer** [comment]
Was there a solution for this?

**Comment by harryseely** [comment]
> Was there a solution for this?

Hi @Palzer I am using wandb (but I think you could apply similar code for other loggers). Here is the solution I came up with. 

`def run_lr_finder(cfg, trainer, model, data_module, wandb_logger):
    # Implement auto lr finder
    tuner = Tuner(trainer)
    model.hparams['lr'] = cfg['lr']  # Ensure model has lr hyperparameter so it can be updated
    lr_finder = tuner.lr_find(model, datamodule=data_module)
    # Only log new lr on main device (0)
    if torch.cuda.current_device() == 0:
        new_lr = lr_finder.suggestion()
        print(f"LR finder identified:\n{new_lr}\nas candidate learning rate")
        fig = lr_finder.plot(suggest=True)
        plot_fpath = "temp/lr_finder_plot.jpg"
        plt.savefig(plot_fpath)
        fig.show()

        # Save plot as a artifact in wandb
        if wandb_logger is False:
            pass
        else:
            wandb_logger.log_image(key="lr_finder_plot", images=[plot_fpath])

            # Set new lr hp in wandb config
            wandb.config.update({"lr": new_lr}, allow_val_change=True)`

**Comment by Ritesh313** [comment]
This still doesn't update automatically when using comet logger. I have learning_rate parameter in the Module definition, it gets updated in the object, but comet still shows the initial learning rate that was passed when creating that object.

```python
# Class definition:
class RGB_Main_Module(L.LightningModule):
    def __init__(self, model, learning_rate):
        ...
        self.save_hyperparameters(ignore=['model'])
    ....

# creating an object
m = RGB_Main_Module(model=model, learning_rate = 0.0001)
tuner = Tuner(trainer=trainer)
lr_finder = tuner.lr_find(model=m, datamodule=data_module, num_training=1000)
```

Should I use `save_hyperparmeters` somewhere else or can i use some other command to update the learning_rate value on comet? Currently I'm just logging the new learning_rate explicitly under another name.

**Comment by SkafteNicki** [comment]
Hi, thanks for raising this issue. I just wanted to confirm that it is indeed correct that hyper parameters are logged before the attributes are updated by the learning rate finder. The precise place in the code is here:
https://github.com/Lightning-AI/pytorch-lightning/blob/cdc0db4898b8cad6dbc8611b8e722280bdf79ec6/src/lightning/pytorch/trainer/trainer.py#L1001-L1002
which happens before learning rate finder starts. The good news is there is way to log the updated learning rate regardless of what logger you may be using. All loggers has the method `log_hyperparams` which is what lightning calls internally in the above mention lines of code. 
Thus, you can just call this again after the learning rate finder is done to re-log the learning rate. A example script:
```python
import tempfile
import torch
import lightning.pytorch as pl
from lightning.pytorch import Trainer
from lightning.pytorch.loggers import CSVLogger
from lightning.pytorch.callbacks import LearningRateFinder
from torch.utils.data import DataLoader, TensorDataset


class SimpleModel(pl.LightningModule):
    def __init__(self, learning_rate=0.5):
        super().__init__()
        self.save_hyperparameters()
        self.layer = torch.nn.Linear(10, 1)
    
    def forward(self, x):
        return self.layer(x)
    
    def training_step(self, batch, batch_idx):
        x, y = batch
        return torch.nn.functional.mse_loss(self(x).squeeze(), y)
    
    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)

with tempfile.TemporaryDirectory() as temp_dir:
    # Create sample data
    torch.manual_seed(42)
    X = torch.randn(1000, 10)
    y = torch.sum(X[:, :3], dim=1) + 0.1 * torch.randn(1000)
    train_loader = DataLoader(TensorDataset(X, y), batch_size=64)
    
    # Model with initial learning rate (will be auto-tuned)
    initial_lr = 0.5
    model = SimpleModel(learning_rate=initial_lr)
    
    # Setup trainer with LearningRateFinder callback
    logger = CSVLogger(temp_dir, name="demo")
    trainer = Trainer(
        max_epochs=2,
        logger=logger,
        callbacks=[LearningRateFinder(min_lr=1e-6, max_lr=1e-1)],
        enable_progress_bar=False,
        enable_model_summary=False
    )

    # Train with automatic learning rate finding
    trainer.fit(model, train_loader)
    trainer.logger.log_hyperparams(model.hparams)  # re-log the hyperparameters
```
(this example uses the `LearningRateFinder` callback but the same "trick" applies to calling the `Tuner.lr_find` method)
For completeness, this behavior is controlled by this argument in the trainer:
https://github.com/Lightning-AI/pytorch-lightning/blob/e1e2534d32cad1d2bad87f5b03da857108773bba/src/lightning/pytorch/trainer/trainer.py#L133
which can be disabled for more control on when logging of hyperparameters happens. 

Closing issue, but feel free to ping me and reopen if necessary.
